--- a/RtAudio.cpp	2024-02-11 15:05:31
+++ b/RtAudio.cpp	2024-02-11 15:07:49
@@ -1782,7 +1782,7 @@
   stream_.deviceFormat[mode] = RTAUDIO_FLOAT32;
 
   if ( streamCount == 1 )
-    stream_.nDeviceChannels[mode] = description.mChannelsPerFrame;
+    stream_.nDeviceChannels[mode] = streamChannels;
   else // multiple streams
     stream_.nDeviceChannels[mode] = channels;
   stream_.nUserChannels[mode] = channels;
